tests/server/api/ui/test_task_runs.py

```diff
@@ -128,3 +128,189 @@ async def test_returns_completed_and_failed_counts(
             TaskRunCount(completed=2, failed=3),
             TaskRunCount(completed=2, failed=2),
         ]
+
+
+class TestReadTaskRunCountsByState:
+    @pytest.fixture
+    def url(self) -> str:
+        return "/ui/task_runs/count"
+
+    @pytest.fixture
+    async def create_flow_runs(
+        self,
+        session: AsyncSession,
+        flow,
+    ):
+        run_1 = await models.flow_runs.create_flow_run(
+            session=session,
+            flow_run=actions.FlowRunCreate(
+                flow_id=flow.id,
+                state=states.Completed(),
+            ),
+        )
+
+        run_2 = await models.flow_runs.create_flow_run(
+            session=session,
+            flow_run=actions.FlowRunCreate(
+                flow_id=flow.id,
+                state=states.Failed(),
+            ),
+        )
+
+        run_3 = await models.flow_runs.create_flow_run(
+            session=session,
+            flow_run=actions.FlowRunCreate(
+                flow_id=flow.id,
+                state=states.Pending(),
+            ),
+        )
+
+        await session.commit()
+
+        return [run_1, run_2, run_3]
+
+    @pytest.fixture
+    async def create_task_runs(
+        self,
+        session: AsyncSession,
+        flow_run,
+        create_flow_runs,
+    ):
+        task_runs_per_flow_run = 27
+        now = cast(DateTimeTZ, pendulum.datetime(2023, 6, 1, 18, tz="UTC"))
```

Comment:
Do you think it would make sense to add some crazy high numbers here to test performance for this? Like hundreds/thousands of task runs?

Reply:
Hm generally I don't think unit tests are the appropriate place for that, but I don't really have a good idea of where we might do that otherwise. Are we doing that elsewhere that you're aware of?

Reply:
> Are we doing that elsewhere that you're aware of?
No, I don't think so. It just seems like it could be quite slow in some cases maybe.

In the issue it notes:
> Users can see a list of task runs on a flow run's page (in the Task Runs tab). It is not possible to see how many tasks are in each possible state. This becomes problematic when there are 100s or 1000s of tasks. 

Since the use case here is "when there are 100s or 1000s of tasks", and I just know that sometimes things have been slow to load before in the UI when it comes to loading runs in certain states (e.g. loading flow runs on the Deployments page when it has thousands of them in a terminal state). I'll be curious to see how it shakes out and if it's something where we end up introducing a setting that users could disable this if needed. I'm probably just too paranoid, ignore me 😅

Reply:
No it's definitely worth considering, I just don't have a great answer 😅

Reply:
Makes sense, thank @bunchesofdonald!